python/python/bystro/tada_mixture/mixture_tada_em.py

```diff
@@ -0,0 +1,394 @@
+        td = self.training_options
+        K = self.K
+
+        N, p = X.shape
+
+        # Initialize variables
+        if Lamb_init is None:
+            model = GaussianMixture(K)
+            model.fit(X)
+            Lambda_ = model.means_
+        else:
+            Lambda_ = Lamb_init
+
+        rng = np.random.default_rng(2021)
+        pi_ = rng.dirichlet(np.ones(K)) if pi_init is None else pi_init
+
+        myrange = trange if progress_bar else range
+
+        for i in myrange(td["n_iterations"]):
+            # E step compute most likely categories
+            Zprobs = np.zeros((N, K))
+            for j in range(K):
+                log_pmf = st.poisson.logpmf(X, Lambda_[j])
+                log_pmf_sum = np.sum(log_pmf, axis=1)
+                Zprobs[:, j] = log_pmf_sum + np.log(pi_[j])
+            Ez = np.argmax(Zprobs, axis=1)
+
+            # M step, maximize parameters
+            for j in range(K):
+                pi_[j] = np.mean(Ez == j)
+                Lambda_[j] = np.mean(X[Ez == j], axis=0)
+
+            eps = 0.001
+            pi_ = pi_ * (1 - K * eps) + np.ones(K) * eps
+
+        self.Lambda = Lambda_
+        self.pi = pi_
+        return self
+
+    def predict(self, data):
+        """
+        This predicts the latent cluster assignments given a fitted model
+
+        Parameters
+        ----------
+        data : np.array-like,shape=(N_variants,n_phenotypes)
+            The data of counts, variants x phenotypes
+
+        Returns
+        -------
+        z_hat : np.array-like,shape=(N_samples,)
+            The cluster identities
+        """
+        log_proba = self.predict_logproba(data)
+        z_hat = np.argmax(log_proba, axis=1)
+        return z_hat
+
+    def predict_logproba(self, data):
+        """
+        Predicts the log probability a datapoint being assigned a specific
+        cluster.
+
+        Parameters
+        ----------
+        data : np.array-like,shape=(N_variants,n_phenotypes)
+            The data of counts, variants x phenotypes
+
+        Returns
+        -------
+        z_hat : np.array-like,shape=(N_samples,)
+            The cluster identities
+        """
+        N = data.shape[0]
+        log_proba = np.zeros((N, self.K))
+        for i in range(N):
+            for k in range(self.K):
+                log_proba[i, k] = np.sum(
+                    st.poisson.logpmf(data[i], self.Lambda[k])
+                )
+        return log_proba
+
+    def _fill_training_options(self, training_options):
+        """
+        This fills any relevant parameters for the learning algorithm
+
+        Parameters
+        ----------
+        training_options : dict
+
+        Returns
+        -------
+        tops : dict
+        """
+        default_options = {"n_iterations": 100}
+        tops = deepcopy(default_options)
+        tops.update(training_options)
+        return tops
+
+
+class MVTadaZPoissonEM(object):
+    def __init__(self, K=4, training_options={}):
```

Comment:
`training_options={}` is a bit of a footgun / language wart in python, because if it's modified in one object it will be modified by default in all others, e.g.:

```python
class Foo:
    def __init__(self, my_dict={}):
        self.my_dict = my_dict


a = Foo()
a.my_dict["hello"] = "world"


b = Foo()
print(b.my_dict)
# {'hello': 'world'}
```

a safer way would be to write something like:
```python
def __init__(training_options=None):
    if training_options is None:
        training_options = {}
```

I'm not sure if this actually causes a problem in this class at the moment, but avoiding initializing vars to mutable collections like lists or dicts will help to keep it safe in the future.

Reply:
Yeah I'll change that. I have no clue why that happens, that's terrible.